packages/datadog-plugin-sharedb/src/index.js

```diff
@@ -0,0 +1,100 @@
+'use strict';
+
+const MessagesAwaitingResponse = new WeakMap();
+const READABLE_ACTION_NAMES = {
+  hs: 'handshake',
+  qf: 'query-fetch',
+  qs: 'query-subscribe',
+  qu: 'query-unsubscribe',
+  bf: 'bulk-fetch',
+  bs: 'bulk-subscribe',
+  bu: 'bulk-unsubscribe',
+  f: 'fetch',
+  s: 'subscribe',
+  u: 'unsubscribe',
+  op: 'op',
+  nf: 'snapshot-fetch',
+  nt: 'snapshot-fetch-by-ts',
+  p: 'presence-broadcast',
+  pr: 'presence-request',
+  ps: 'presence-subscribe',
+  pu: 'presence-unsubscribe'
+};
+
+function createTraceName(action, collection) {
+  let actionName = READABLE_ACTION_NAMES[action];
+  if (actionName === undefined) {
+    actionName = action;
+  }
+  let traceName = 'sharedb-request/' + actionName;
+  if (collection) {
+    traceName += '/' + collection;
+  }
+  return traceName;
+}
+
+function createWrapHandle(tracer, config) { // called once
+  return function wrapTrigger(triggerFn) { // called once
+    return function handleMessageWithTrace(action, agent, triggerContext, callback) { // called for each trigger
+      /**
+       * What we're doing here is tying ourselves into the ShareDB Backend middleware.
+       * This allows us to create traces for all events that have triggers, like receiving a message and replying
+       * to it. The benefit of doing this over wrapping the connection class is that both the receive and reply
+       * triggers have access to a reference of the original request object. This allows us to use a WeakMap to
+       * store the span call backs to help prevent memory leaks.
+       *
+       */
+      switch (action) {
+        case 'receive':
+          if (triggerContext.data && triggerContext.data.a) {
+            const scope = tracer.scope();
+            const childOf = scope.active();
+            return triggerFn.call(this, action, agent, triggerContext, function wrappedCallback(err) {
+              tracer.trace(
+                createTraceName(triggerContext.data.a, triggerContext.data.c),
+                { childOf },
+                (span, spanDoneCb) => {
+                  if (config.hooks && config.hooks.receive) {
+                    config.hooks.receive(span, agent, triggerContext);
```

Comment:
Hooks should be called at the last minute right before `span.finish()` is called.

Reply:
I think that we should be able to handle different types of hooks, right? I think what you're suggesting is missing is a "reply" hook?

For example at Lever we need the hook when the span is created and before the next set of operations happen.